### DISCUSSION TITLE
How to specify autoincremental IDs supporting SqLite and MsSql?

### DISCUSSION THREAD

**Discussion by Slesa** [fluentmigrator/fluentmigrator#1295] [Q&A] [answered]
When developping with MS Sql LocalDB, I've started with a migration
```
Create.Table("Bons").WithColumn("Id").AsInt64().PrimaryKey().Identity()
```
which works perfectly with new rows - I was happy. 
Given this version to QS, they said "But hey, there is no generated data at all?" They are using SqLite. After some trouble with Guids, I found out that only
```
Create.Table("Bons").WithColumn("Id").AsInt64().PrimaryKey()
```
works here. Docs of SqLite say: try to avoid autoinc!

Knowing, that different creation for different DB types is possible, I started with
```
ICreateTableColumnAsTypeSyntax CreateIdRow(string tableName) {
  var result = IfDatabase("SQLite")
               .Create.Table(tableName).WithColumn("Id").AsInt64().PrimaryKey();
  if( result.IsEmpty() ) // ???
```

There seems to be no proper way to add an expression to the create statement depending on the DB type, doesn't it?
The only way is to put the if condition all around each and every table creation:
```
IfDatabase("SqlServer", "Postgres").
  Create.Table("Bons")
   .WithColumn("Id").AsInt64().PrimaryKey().Identity()
   . ... name ...
   . ... description ...
   . ... date/time ...
   . ... waiter ...
   . ... all other fields ...
IfDatabase("SqLite").
  Create.Table("Bons")
   .WithColumn("Id").AsInt64().PrimaryKey()
   . ... all those fields again ...
```
Is that true or am I missing something?

**Comment by jzabroski** [comment]
@Slesa 

> There seems to be no proper way to add an expression to the create statement depending on the DB type, doesn't it?

Unfortunately, this doesn't exist.  There are explanations for why it doesn't exist, and some discussion in private between myself and another person on how to rectify this.

There is one workaround to your problem: Use an extension method.

```c#
public static IMigrationExtension
{
  public static ICreateTableColumnOptionOrWithColumnSyntax /* TODO Verify this pseudo-code is the correct interface */
     CreateBonsTable(this IMigration migration, bool withIdentity)
  {
    var expr = this.Create.Table("Bons")
      .WithColumn("Id").AsInt64().PrimaryKey()
    if (withIdentity) expr.Identity()
      . ... name ...
      . ... description ...
      . ... date/time ...
      . ... waiter ...
      . ... all other fields ...
  }
}
```

then you can write:

```c#
IfDatabase("SqlServer", "Postgres").CreateBonsTable(withIdentity: true);
IfDatabase("SqlLite").CreateBonsTable(withIdentity: false);
```

It's not perfect but it does solve the problem relating to the capturing of Abstract Syntax Tree root, and is basically the only workaround with an AST based approach.  I've been discussing with someone a different approach, but its not clear to me the other approach is superior (it has drawbacks such as requiring dynamic runtime type checks).

**Comment by jzabroski** [comment]
Edited original pseudo-code above. Please refresh.

**Comment by jzabroski** [comment]
BTW, as a best practice, you should explicitly specify Nullable/NotNullable, as FM defaults to NotNullable.  

For more examples on using extension methods, see the examples folder: https://github.com/fluentmigrator/fluentmigrator/blob/e82aafa20e6dbe3cefa221303fe23cf8bf59fffd/samples/FluentMigrator.Example.Migrations/MigrationExtensions.cs#L25

**Comment by Slesa** [comment]
Mmh... the fluent base is killing everything. Either IfDatabase or Create, one of them is the root. 
If there would be an extension 
```
ICreateTableColumnOption .OnlyWhen( Func<bool> ), 
```
wouldn't that help?

**Comment by Slesa** [comment]
It feels strange to give an incremental primary key the option 'NotNullable'...

Anyway, I think my real problem is that the SqLite provider puts the autoincremental option to the expression. Is it imaginable to have a second extension for SqLite (and I think MySql, too? Not sure) WithAutoincrement()?

**Comment by Slesa** [comment, marked as answer]
This is the shortest code I've found. I will need that for each table with auto IDs - so almost all.

To be repeated for each table:
```
        public override void Up()
        {
            IfDatabase("SqlServer", "Postgres").CreateAutoIncrementIdTable("Bons")
                .WithBonFields();
            IfDatabase("Sqlite").CreateIdTable("Bons")
                .WithBonFields();
            ...
```

Extension methods for all tables to declare the columns:
```
        public static ICreateTableColumnOptionOrWithColumnSyntax WithBonFields(
            this ICreateTableWithColumnSyntax tableWithColumnSyntax)
        {
            return tableWithColumnSyntax
               ... all column definitions 
               ...
```

Extension for table creation itself:
```
        public static ICreateTableColumnOptionOrWithColumnSyntax WithAutoIncrementId(
            this ICreateTableWithColumnSyntax tableWithColumnSyntax)
        {
            return tableWithColumnSyntax
                .WithColumn("Id")
                .AsInt64()
                .NotNullable()
                .PrimaryKey()
                .Identity();
        }

        // For SqLite where AutoInc leads to trouble
        public static ICreateTableColumnOptionOrWithColumnSyntax WithIdColumn(
            this ICreateTableWithColumnSyntax tableWithColumnSyntax)
        {
            return tableWithColumnSyntax
                .WithColumn("Id")
                .AsInt64()
                .NotNullable()
                .PrimaryKey();
        }

        public static ICreateTableColumnOptionOrWithColumnSyntax CreateAutoIncrementIdTable(
            this IIfDatabaseExpressionRoot migration, string name)
        {
            return migration.Create.Table(name)
                .WithAutoIncrementId();
        }
        
        public static ICreateTableColumnOptionOrWithColumnSyntax CreateIdTable(
            this IIfDatabaseExpressionRoot migration, string name)
        {
            return migration.Create.Table(name)
                .WithIdColumn();
        }
```
Maybe there are ways to use your first suggestion, but it is difficult to find the right interface to extend (it is not IMigration, maybe IIfDatabaseExpressionRoot) and it only saves a few lines.

**Comment by jzabroski** [comment]
> Maybe there are ways to use your first suggestion, but it is difficult to find the right interface to extend (it is not IMigration, maybe IIfDatabaseExpressionRoot) and it only saves a few lines.

I dont know if this helps, but I had a typo in my original example.  It should be `var expr = migration.Create.Table`.  The basic trick is of using an extension method is to capture the "this" variable to avoid referring to the wrong "this", since the value of "this" changes at run-time (literally, its internal state accumulator of the runtime AST changes depending on which IfDatabase condition is true).

> Mmh... the fluent base is killing everything. Either IfDatabase or Create, one of them is the root.
> If there would be an extension
> 
> ```
> ICreateTableColumnOption .OnlyWhen( Func<bool> ), 
> ```
> 
> wouldn't that help?

Sort of.  Part of the problem with allowing arbitrary code to be interpreted is that it might violate the underlying database's transaction semantics.  I can go into detail here, but the basic idea is that some expressions in various SQL dialects must be the only statement in a batch (e.g., CREATE VIEW) and only dialects with Procedural SQL extensions have workarounds.  Not sure if you've ever heard of the Joel Spolsky "Leaky Abstraction" argument when designing an API/software sytem, but I would like FluentMigrator API to leak the least amount of underlying details.

I have started sketching out ideas on how to simplify this using "F-Bound polymorphism style", and somebody in private has suggested "Tagless-Final style" API.  The problem with Tagless-Final style is that it basically makes it very hard to check the properties of things like transactional consistency and that all commands are closed under composition.  Not impossible, just hard.

One alternative to OnlyWhen you propose would be: https://github.com/fluentmigrator/fluentmigrator/issues/880

However, the reason all these are "workarounds" comes back to whether you can query metadata as the runner executes, and react to that metadata.  Due to reasons around dealing with transactional consistency, things like #1263 are problems.

**Reply by jzabroski** [reply]
Actually, re-reading #880, I propose one solution that could work for you (I can't remember if I actually got this idea to work).  https://github.com/fluentmigrator/fluentmigrator/issues/880#issuecomment-395789400
